probe-rs-tools/src/bin/probe-rs/cmd/run.rs

```diff
@@ -255,39 +264,164 @@ impl Cmd {
 #[derive(PartialEq)]
 enum RunMode {
     Normal,
-    Test,
+    Test(EmbeddedTestElfInfo),
 }
 
-fn elf_contains_test(path: &Path) -> anyhow::Result<bool> {
-    let mut file = File::open(path).map_err(FileDownloadError::IO)?;
+#[derive(Debug, PartialEq)]
+pub struct EmbeddedTestElfInfo {
+    /// Protocol version used between embedded-test (on the target) and probe-rs
+    pub version: u32,
+    /// Tests found in the elf.
+    pub tests: Vec<Test>,
+}
 
-    let mut buffer = Vec::new();
-    file.read_to_end(&mut buffer)?;
+struct ElfReader<'a> {
+    buffer: &'a [u8],
+    elf: Elf<'a>,
+}
 
-    let contains = match goblin::elf::Elf::parse(buffer.as_slice()) {
-        Ok(elf) if elf.syms.is_empty() => {
+impl<'a> ElfReader<'a> {
+    fn decode(&self) -> anyhow::Result<Option<EmbeddedTestElfInfo>> {
+        if self.elf.syms.is_empty() {
             tracing::debug!("No Symbols in ELF");
-            false
+            return Ok(None);
         }
-        Ok(elf) => elf
-            .syms
+
+        // Find our custom .embedded_test section which contains version info and the tests
+        let Some((et_section_index, et_section)) = self
+            .elf
+            .section_headers
             .iter()
-            .any(|sym| elf.strtab.get_at(sym.st_name) == Some("EMBEDDED_TEST_VERSION")),
-        Err(_) => {
-            tracing::debug!("Failed to parse ELF file");
-            false
+            .enumerate()
+            .find(|(_idx, hdr)| self.elf.shdr_strtab.get_at(hdr.sh_name) == Some(".embedded_test"))
+        else {
+            tracing::debug!("No .embedded_test linker section in ELF");
+            return Ok(None);
+        };
+
+        // Iterate over all symbols of our section
+        // Symbols we look for:
+        // * Version Info: 4 byte symbol with name "EMBEDDED_TEST_VERSION"
+        // * Testcase: 12 byte symbol, where the name is an escaped json_object.
+
+        let mut version = None;
+        let mut tests = vec![];
+        for sym in self.elf.syms.iter() {
+            if sym.st_bind() == STB_GLOBAL
+                && sym.st_type() == STT_OBJECT
+                && sym.st_shndx == et_section_index
+            {
+                let sym_name = self
+                    .elf
+                    .strtab
+                    .get_at(sym.st_name)
+                    .ok_or(anyhow!("No name for symbol {:?}", sym))?;
+
+                if sym.st_size == 4 && sym_name == "EMBEDDED_TEST_VERSION" {
+                    version = Some(
```

Comment:
You should check the version here, while you are parsing the .elf. Otherwise, how do you know that the data you've decoded is valid?

Reply:
Good point. fixed.